### PR TITLE
Add nil type descriptor support

### LINKED ISSUE #22067: Add nil type

### PR DESCRIPTION
**Pull request by rdulmina**
## Purpose
Add nil type descriptor support to new parser with test cases and recovery logic.

Fixes #22067

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [x] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [x] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: compiler/ballerina-parser/src/test/java/io/ballerinalang/compiler/parser/test/TestParser.java [hunk 1]
```diff
@@ -36,7 +36,7 @@
     private static final PrintStream OUT = System.out;
 
     public static void main(String[] args) throws IOException {
-        String path = "src/test/resources/test1.bal";
+        String path = "compiler/ballerina-parser/src/test/resources/test1.bal";
```
Comment (SupunS) [inline compiler/ballerina-parser/src/test/java/io/ballerinalang/compiler/parser/test/TestParser.java:39]: need this change?

File: compiler/ballerina-parser/src/test/resources/types/simple-types/simple_types_source_01.bal [hunk 2]
```diff
@@ -12,4 +12,5 @@ public function foo() {
     service k;
     var l;
     never m;
+    () n;
```
Comment (SupunS) [inline compiler/ballerina-parser/src/test/resources/types/simple-types/simple_types_source_01.bal:15]: Shall we add some negative test as well? such as missing one of the parenthesis, additional tokens between the parenthesis, etc?

File: compiler/ballerina-parser/src/test/java/io/ballerinalang/compiler/parser/test/ParserTestUtils.java [hunk 3]
```diff
@@ -489,6 +489,8 @@ private static SyntaxKind getNodeKind(String kind) {
                 return SyntaxKind.BREAK_STATEMENT;
             case "RETURN_STATEMENT":
                 return SyntaxKind.RETURN_STATEMENT;
+            case "NIL_TYPE":
+                return SyntaxKind.NIL_TYPE;
```
Comment (SupunS) [inline compiler/ballerina-parser/src/test/java/io/ballerinalang/compiler/parser/test/ParserTestUtils.java:493]: shall we move this to the 'Others' section in the switch case..?

File: compiler/ballerina-parser/src/test/java/io/ballerinalang/compiler/parser/test/TestParser.java (hunk 1 above)
Comment (rdulmina) [reply compiler/ballerina-parser/src/test/java/io/ballerinalang/compiler/parser/test/TestParser.java:39]: No, I'll correct that

File: compiler/ballerina-parser/src/test/resources/types/simple-types/simple_types_source_01.bal (hunk 2 above)
Comment (rdulmina) [reply compiler/ballerina-parser/src/test/resources/types/simple-types/simple_types_source_01.bal:15]: sure

File: compiler/ballerina-parser/src/test/java/io/ballerinalang/compiler/parser/test/ParserTestUtils.java (hunk 3 above)
Comment (rdulmina) [reply compiler/ballerina-parser/src/test/java/io/ballerinalang/compiler/parser/test/ParserTestUtils.java:493]: Sure @SupunS

File: compiler/ballerina-parser/src/test/resources/types/simple-types/simple_types_source_01.bal (hunk 2 above)
Comment (rdulmina) [reply compiler/ballerina-parser/src/test/resources/types/simple-types/simple_types_source_01.bal]: @SupunS I have added more test cases.

**Pull request merged**